pkg/apis/meta.go

```diff
@@ -0,0 +1,303 @@
+	obj runtime.Object
+}
+
+func MetaAccessor(obj runtime.Object) GrafanaResourceMetaAccessor {
+	return &grafanaResourceMetaAccessor{obj}
+}
+
+func (m *grafanaResourceMetaAccessor) set(key string, val string) {
+	acc, err := meta.Accessor(m.obj)
+	if err != nil {
+		return
+	}
+
+	anno := acc.GetAnnotations()
+	if val == "" {
+		if anno != nil {
+			delete(anno, key)
+		}
+	} else {
+		if anno == nil {
+			anno = make(map[string]string)
+		}
+		anno[key] = val
+	}
+	acc.SetAnnotations(anno)
+}
+
+func (m *grafanaResourceMetaAccessor) get(key string) string {
+	acc, err := meta.Accessor(m.obj)
+	if err != nil {
+		return ""
+	}
+
+	return acc.GetAnnotations()[key]
+}
+
+func (m *grafanaResourceMetaAccessor) GetUpdatedTimestamp() (*time.Time, error) {
+	acc, err := meta.Accessor(m.obj)
+	if err != nil {
+		return nil, err
+	}
+
+	v, ok := acc.GetAnnotations()[AnnoKeyUpdatedTimestamp]
+	if !ok || v == "" {
+		return nil, nil
+	}
+	t, err := time.Parse(time.RFC3339, v)
+	if err != nil {
+		return nil, fmt.Errorf("invalid updated timestamp: %s", err.Error())
+	}
+	return &t, nil
+}
+
+func (m *grafanaResourceMetaAccessor) SetUpdatedTimestampMillis(v int64) {
+	if v > 0 {
+		t := time.UnixMilli(v)
+		m.SetUpdatedTimestamp(&t)
+	} else {
+		m.SetUpdatedTimestamp(nil)
+	}
+}
+
+func (m *grafanaResourceMetaAccessor) SetUpdatedTimestamp(v *time.Time) {
+	txt := ""
+	if v != nil {
+		txt = v.UTC().Format(time.RFC3339)
+	}
+	m.set(AnnoKeyUpdatedTimestamp, txt)
+}
+
+func (m *grafanaResourceMetaAccessor) GetCreatedBy() string {
+	return m.get(AnnoKeyCreatedBy)
+}
+
+func (m *grafanaResourceMetaAccessor) SetCreatedBy(user string) {
+	m.set(AnnoKeyCreatedBy, user)
+}
+
+func (m *grafanaResourceMetaAccessor) GetUpdatedBy() string {
+	return m.get(AnnoKeyUpdatedBy)
+}
+
+func (m *grafanaResourceMetaAccessor) SetUpdatedBy(user string) {
+	m.set(AnnoKeyUpdatedBy, user)
+}
+
+func (m *grafanaResourceMetaAccessor) GetFolder() string {
+	return m.get(AnnoKeyFolder)
+}
+
+func (m *grafanaResourceMetaAccessor) SetFolder(uid string) {
+	m.set(AnnoKeyFolder, uid)
+}
+
+func (m *grafanaResourceMetaAccessor) GetSlug() string {
+	return m.get(AnnoKeySlug)
+}
+
+func (m *grafanaResourceMetaAccessor) SetSlug(v string) {
+	m.set(AnnoKeySlug, v)
+}
+func (m *grafanaResourceMetaAccessor) GetTitle() string {
+	// look for Spec.Title or Spec.Name
+	r := reflect.ValueOf(m.obj)
+	if r.Kind() == reflect.Ptr || r.Kind() == reflect.Interface {
+		r = r.Elem()
+	}
+	if r.Kind() == reflect.Struct {
+		spec := r.FieldByName("Spec")
+		if spec.Kind() == reflect.Struct {
+			title := spec.FieldByName("Title")
+			if title.IsValid() && title.Kind() == reflect.String {
+				return title.String()
+			}
+			name := spec.FieldByName("Name")
+			if name.IsValid() && name.Kind() == reflect.String {
+				return name.String()
+			}
+		}
+	}
+
+	// fall back to annotation
+	return m.get(AnnoKeyTitle)
+}
+
+func (m *grafanaResourceMetaAccessor) SetTitle(v string) {
```

Comment:
i don't think we want this one -- if it is really necessary then maybe, but feels to dangerous